test/e2e/framework/framework.go

```diff
@@ -279,3 +279,26 @@ func UpdateDeployment(kubeClientSet kubernetes.Interface, namespace string, name
 
 	return nil
 }
+
+// UpdateConfigmap updates the ingress-nginx config
+func UpdateConfigmap(k, v string, c kubernetes.Interface) error {
+	By(fmt.Sprintf("updating configuration configmap setting %v to '%v'", k, v))
+	config, err := c.CoreV1().ConfigMaps("ingress-nginx").Get("nginx-configuration", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if config == nil {
+		return fmt.Errorf("config map does not exist")
+	}
+	
+	if config.Data == nil {
+		config.Data = map[string]string{}
+	}
+	config.Data[k] = v
+	_, err = c.CoreV1().ConfigMaps("ingress-nginx").Update(config)
+	if err != nil {
+		return err
+	}
+	time.Sleep(1 * time.Second)
+	return nil
```

Comment:
Could we return a deep copy of the old configMap before the update? This way we can reset the configMap at the end of a test without mutating state

Reply:
Is this pattern used elsewhere in the framework? Returning the previous state from a function that is altering the config map seems a little bit confusing. I think it might be more logical to have a ConfigMapReset function

Reply:
Hmm, I think you're right - it would be cleaner, but my concern is - how would the method know what is the default? The developer might have kubectl-d her own configMap, and I'm not sure assuming that [deploy/configmap.yaml](https://github.com/kubernetes/ingress-nginx/blob/master/test/e2e/wait-for-nginx.sh#L22) has the defaults is the right approach. 

Let me know what you think / if that resonates. When the e2e-tests run, ingress-nginx and everything in it should already be a thing, so I'm not sure at what point we'd capture the initial state

Reply:
> Is this pattern used elsewhere in the framework

Not really, since there aren't any tests that require changing ingress-nginx. For example wherever `updateDeployment` is used - it doesn't need to reset anything as it is only used on the ephemeral EchoDeployment that's created before each test run, and destroyed right after.

Reply:
I get what you mean; default it sort of arbitrary. I think it would be better to deepcopy the config map within the test as a local variable _before_ updating and then reset that way.

Reply:
Another solution could be to have a `AddNginxConfigMapAnnotation` function that would add/update the annotation AND have a `RemoveNginxConfigMapAnnotation`. That way no copying would be needed and annotations could easily be added and revoked as needed.

Reply:
A similar pattern is used in dynamic_configuation.go for `enableDynamicConfiguration` and `disableDynamicConfiguration`

Reply:
Using Add and `RemoveNginxConfigMapAnnotation` might not work as expected - since there might have been a pre-existing value that we've overridden.

> I think it would be better to deepcopy the config map within the test as a local variable before updating and then reset at the end of the test.

I think this approach is pretty solid, and I wound up doing exactly this over here: https://github.com/Shopify/ingress/pull/44/files#diff-8700c46a8379c603b5e41ccbaff6b71eR123 (lines 123 - 129)

and here: 

https://github.com/Shopify/ingress/pull/44/files#diff-8700c46a8379c603b5e41ccbaff6b71eR153

I realized that those few lines would be repeated all over when we'd want to rollback - and returning the old configmap would avoid the trouble of having to rollback appropriately based on whether there _was_ a pre-existing value - or if there wasn't one